sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,239 @@
+                scope.Failed(requestFailedException);
+                throw requestFailedException;
+            }
+        }
+    }
+
+#pragma warning disable SA1402 // File may only contain a single type
+    internal class OperationInternal<TResult, TResponseType> : OperationInternal<TResult>
+#pragma warning restore SA1402 // File may only contain a single type
+    {
+        private readonly IOperation<TResult, TResponseType> _operation;
+
+        public OperationInternal(string operationTypeName,
+            ClientDiagnostics clientDiagnostics,
+            IOperation<TResult, TResponseType> operation,
+            TimeSpan? defaultPollingInterval = null) : base(operationTypeName, clientDiagnostics, defaultPollingInterval)
+        {
+            _operation = operation;
+        }
+
+        protected override async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = ClientDiagnostics.CreateScope($"{OperationTypeName}.UpdateStatus");
+
+            _operation.AddAttributes(scope);
+            scope.Start();
+
+            try
+            {
+                Response<TResponseType> response = async
+                    ? await _operation.GetResponseAsync(cancellationToken).ConfigureAwait(false)
+                    : _operation.GetResponse(cancellationToken);
+
+                Response rawResponse = response.GetRawResponse();
+                RawResponse = rawResponse;
+
+                var status = _operation.UpdateState(response);
+
+                if (status == CompletionStatus.Succeeded)
+                {
+                    Value = _operation.ParseResponse(response);
+                    HasCompleted = true;
+                }
+                else if (status == CompletionStatus.Failed)
+                {
+                    OperationFailedException = _operation.GetOperationFailedException(response);
+                    HasCompleted = true;
+
+                    throw OperationFailedException;
+                }
+
+                return rawResponse;
+            }
+            catch (RequestFailedException e)
+            {
+                scope.Failed(e);
+                throw;
+            }
+            catch (Exception e)
+            {
+                var requestFailedException = new RequestFailedException("Unexpected failure.", e);
+                scope.Failed(requestFailedException);
+                throw requestFailedException;
+            }
+        }
+    }
+
+    internal interface IOperation<TResult>
+    {
+        Task<Response> GetResponseAsync(CancellationToken cancellationToken);
+
+        Response GetResponse(CancellationToken cancellationToken);
+
+        CompletionStatus UpdateState(Response response);
+
+        TResult ParseResponse(Response response);
+
+        RequestFailedException GetOperationFailedException(Response response);
+
+        void AddAttributes(DiagnosticScope scope);
+    }
```

Comment:
The only thing I don't like about using interfaces is that not every method will be used by every client. For instance, `AddAttributes` for diagnostic scopes is rarely used. We could have a separate abstract class, but developers would need to create their own `OperationInternal` concrete implementations in addition to the main Operation class, so it wouldn't feel very practical.

Reply:
What is typically added in AddAttributes? Can they be added once in the ctor?

Reply:
You have a point. I believe all attributes we're adding are immutable.
- **KeyVault.Keys:** key name
- **KeyVault.Secrets:** secret name
- **MixedReality.RemoteRendering:** session ID (need to confirm, but I think it's immutable as well)